File: 0203. Remove Linked List Elements/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeElements(self, head: ListNode, val: int) -> ListNode:
        dummy = ListNode(None)
        dummy.next = head
        pre = dummy
        cur = head
        while (cur):
            if cur.val == val:
                pre.next = cur.next
                if cur == head:
                    head = cur.next
                cur = cur.next
                continue
            cur = cur.next
            pre = pre.next
        return head

File: 0203. Remove Linked List Elements/solution.java
/**
 * My solution of one pass
 *
 * Time: O(n). n is the length of linked list.
 * Space: O(1)
 */
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        if (head == null) return head;
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode cur = head, prev = dummy;
        while (cur != null) {
            if (cur.val == val)
                prev.next = cur.next;
            else
                prev = cur;
            cur = cur.next;
        }
        return dummy.next;
    }
}

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
